drv/lpc55-update-server/src/main.rs

```diff
@@ -573,6 +623,411 @@ impl ServerImpl<'_> {
+        loop {
+            match self.flash.poll_erase_or_program_result() {
+                None => continue,
+                Some(Ok(())) => return Ok(()),
+                Some(Err(_)) => return Err(UpdateError::FlashError),
+            }
+        }
+    }
+
+    fn compare_cache_to_flash(
+        &self,
+        span: &Range<u32>,
+    ) -> Result<(), UpdateError> {
+        // Is there a cached image?
+        // no, return error
+
+        // Lengths are rounded up to a flash page boundary.
+        let clen = self.valid_cache_image_present()?;
+        let flen = self.valid_flash_image_present(span)?;
+        if clen != flen {
+            return Err(UpdateError::ImageMismatch);
+        }
+        // compare flash page to cache
+        let cached =
+            self.fw_cache[0..flen / core::mem::size_of::<u32>()].as_bytes();
+        let mut flash_page = [0u8; BYTES_PER_FLASH_PAGE];
+        for addr in (0..flen).step_by(BYTES_PER_FLASH_PAGE) {
+            let size = if addr + BYTES_PER_FLASH_PAGE > flen {
+                flen - addr
+            } else {
+                BYTES_PER_FLASH_PAGE
+            };
+
+            indirect_flash_read(
+                &self.flash,
+                addr as u32,
+                &mut flash_page[..size],
+            )?;
+            // XXX fw_cache is [u32; FW_CACHE_MAX]
+            if flash_page[0..size] != cached[addr..addr + size] {
+                return Err(UpdateError::ImageMismatch);
+            }
+        }
+        Ok(())
+    }
+
+    // Looking at a region of flash, determine if there is a possible NXP
+    // image programmed. Return the length in bytes of the flash pages
+    // comprising the image including padding to fill to a page boundary.
+    fn valid_flash_image_present(
+        &self,
+        span: &Range<u32>,
+    ) -> Result<usize, UpdateError> {
+        let buf = &mut [0u32; 1];
+        match indirect_flash_read(
+            &self.flash,
+            span.start + LENGTH_OFFSET as u32,
+            buf[..].as_bytes_mut(),
+        ) {
+            Ok(()) => {
+                if let Some(len) = round_up_to_flash_page(buf[0]) {
+                    // The minimum image size should be further constrained
+                    // but this is enough bytes for an NXP header and not
+                    // bigger than the flash slot.
+                    if len as usize <= span.len() && len >= HEADER_OFFSET {
+                        return Ok(len as usize);
+                    }
+                }
+                Err(UpdateError::BadLength)
+            }
+            Err(error) => Err(error),
+        }
+    }
+
+    fn valid_cache_image_present(&self) -> Result<usize, UpdateError> {
+        let len = if let Some(len) = round_up_to_flash_page(
+            self.fw_cache[LENGTH_OFFSET / core::mem::size_of::<u32>()],
+        ) {
+            len
+        } else {
+            return Err(UpdateError::BadLength);
+        };
```

Comment:
style nit: the if-else feels a bit hard to follow due to the weird way `rustfmt` wants to format this,what do you think of:
```suggestion
        let len =  round_up_to_flash_page(
            self.fw_cache[LENGTH_OFFSET / core::mem::size_of::<u32>()],
        ).ok_or(UpdateError::BadLength)?;
```